### PR TITLE
feat(storage): Add ability to restore soft deleted objects

### PR DESCRIPTION
**Pull request by ddelgrosso1**
Fixes: #13575 

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14710)

### CODE REVIEW

**Comment by ddelgrosso1** [comment]
This should be the final piece of #13575 (RestoreObject).

File: google/cloud/storage/client.h [hunk 1]
```diff
@@ -1646,6 +1646,52 @@ class Client {
...
+   *
+   * When a soft-deleted object is restored, a new copy of that object is
+   * created in the same bucket and inherits the same metadata as the
+   * soft-deleted object. The inherited metadata is the metadata that existed
+   * when the original object became soft deleted, with the following
+   * exceptions:
+   *
+   * 1. The createTime of the new object is set to the time at which the
+   * soft-deleted object was restored.
+   * 2. The softDeleteTime and hardDeleteTime values are cleared.
```
Comment (bajajneha27) [inline google/cloud/storage/client.h:1661]: nit: Should the fields `softDeleteTime` and `hardDeleteTime` be snaked_cased ?

File: google/cloud/storage/client.h [hunk 2]
```diff
@@ -1646,6 +1646,52 @@ class Client {
...
+   * @param generation specifies the version of the soft-deleted object to
+   * restore.
+   * @param options a list of optional query parameters and/or request headers.
+   *     Valid types for this operation include `DestinationKmsKeyName`,
+   *      `IfGenerationMatch`, `IfGenerationNotMatch`, `IfMetagenerationMatch`,
+   *      `IfMetagenerationNotMatch`, `Projection`, `CopySourceAcl`
+   *
+   * @return The metadata of the restored object.
+   *
+   * @par Idempotency
```
Comment (bajajneha27) [inline google/cloud/storage/client.h:1679]: What does `@par` mean?

File: google/cloud/storage/internal/connection_impl_object_test.cc [hunk 3]
```diff
@@ -370,6 +370,34 @@ TEST(StorageConnectionImpl, PatchObjectPermanentFailure) {
   EXPECT_THAT(permanent.captured_authority_options(), RetryLoopUsesOptions());
 }
 
+TEST(StorageConnectionImpl, RestoreObjectTooManyFailures) {
```
Comment (bajajneha27) [inline google/cloud/storage/internal/connection_impl_object_test.cc:373]: nit: can we add a happy test case too?

**Review by bajajneha27 — APPROVED** [review]
LGTM from my end.

File: google/cloud/storage/client.h (hunk 1 above)
Comment (ddelgrosso1) [reply google/cloud/storage/client.h:1661]: I don't think so. This is how the JSON API would return these fields (and how they are documented).

File: google/cloud/storage/client.h (hunk 2 above)
Comment (ddelgrosso1) [reply google/cloud/storage/client.h:1679]: https://www.doxygen.nl/manual/commands.html#cmdpar

File: google/cloud/storage/internal/connection_impl_object_test.cc (hunk 3 above)
Comment (ddelgrosso1) [reply google/cloud/storage/internal/connection_impl_object_test.cc:373]: So looking at the other tests in this file it only tests the failure cases for connections. Just for the sake of continuity I'm not sure we want to add others here.

**Review by scotthart — APPROVED** [review]
> Reviewed 10 of 32 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14710)* status: 10 of 32 files reviewed, 4 unresolved discussions (waiting on @bajajneha27 and @ddelgrosso1)

___
*[`google/cloud/storage/internal/object_requests.h` line 391 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14710#-O7-qB3c4Y815M8BJFF8:-O7-qB3c4Y815M8BJFF9:ba7bqsd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b6214b9d20a960fa3fa7a3cbaa43bdef1437df95/google/cloud/storage/internal/object_requests.h#L391)):*
> ```objc
>  public:
>   RestoreObjectRequest() = default;
>   explicit RestoreObjectRequest(std::string bucket_name,
> ```

`explicit` is unnecessary here as the constructor has more than one parameter.

**Pull request merged**